sql/sql_admission_control.h

```diff
@@ -0,0 +1,356 @@
+     PSI_DOCUMENT_ME}};
+
+static void init_st_ac_node_keys() {
+  const char *const category = "sql";
+  int count;
+
+  count = static_cast<int>(array_elements(all_ac_node_mutexes));
+  mysql_mutex_register(category, all_ac_node_mutexes, count);
+
+  count = static_cast<int>(array_elements(all_ac_node_conds));
+  mysql_cond_register(category, all_ac_node_conds, count);
+}
+#endif
+
+struct st_ac_node {
+  mysql_mutex_t lock;
+  mysql_cond_t cond;
+  st_ac_node() {
+#ifdef HAVE_PSI_INTERFACE
+    init_st_ac_node_keys();
+#endif
+    mysql_mutex_init(key_ac_node_lock, &lock, MY_MUTEX_INIT_FAST);
+    mysql_cond_init(key_ac_node_cond, &cond);
+  }
+
+  ~st_ac_node() {
+    mysql_mutex_destroy(&lock);
+    mysql_cond_destroy(&cond);
+  }
+};
+
+/**
+  Class used in admission control.
+
+  Every entity (database or table or user name) will have this
+  object created and stored in the global map AC::ac_map.
+
+*/
+class Ac_info {
+  friend class AC;
+#ifdef HAVE_PSI_INTERFACE
+  PSI_mutex_key key_lock;
+  PSI_mutex_info key_lock_info[1] = {{&key_lock, "Ac_info::lock", 0}};
+#endif
+  // Queue to track the running and waiting threads.
+  std::deque<std::shared_ptr<st_ac_node>> queue;
+  std::atomic<unsigned long> waiting_queries;
+  // Protects the queue.
+  mysql_mutex_t lock;
+
+ public:
+  Ac_info() {
+#ifdef HAVE_PSI_INTERFACE
+    mysql_mutex_register("sql", key_lock_info, array_elements(key_lock_info));
+#endif
+    mysql_mutex_init(key_lock, &lock, MY_MUTEX_INIT_FAST);
+    waiting_queries = 0;
+  }
+  ~Ac_info() { mysql_mutex_destroy(&lock); }
+  // Disable copy constructor.
+  Ac_info(const Ac_info &) = delete;
+  Ac_info &operator=(const Ac_info &) = delete;
+};
+
+/**
+  Global class used to enforce per admission control limits.
+*/
+class AC {
+  // This map is protected by the rwlock LOCK_ac.
+  std::unordered_map<std::string, std::shared_ptr<Ac_info>> ac_map;
+  // Variables to track global limits
+  ulong max_running_queries, max_waiting_queries;
+  /**
+    Protects ac_map and max_running_queries/max_waiting_queries.
+
+    Locking order followed is LOCK_ac, Ac_info::lock, st_ac_node::lock.
+  */
+  mysql_rwlock_t LOCK_ac;
+#ifdef HAVE_PSI_INTERFACE
+  PSI_rwlock_key key_rwlock_LOCK_ac;
+  PSI_rwlock_info key_rwlock_LOCK_ac_info[1] = {
+      {&key_rwlock_LOCK_ac, "AC::rwlock", 0}};
+#endif
+
+  std::atomic<ulonglong> total_aborted_queries;
+
+ public:
+  AC() {
+#ifdef HAVE_PSI_INTERFACE
+    mysql_rwlock_register("sql", key_rwlock_LOCK_ac_info,
+                          array_elements(key_rwlock_LOCK_ac_info));
+#endif
+    mysql_rwlock_init(key_rwlock_LOCK_ac, &LOCK_ac);
+    max_running_queries = 0;
+    max_waiting_queries = 0;
+    total_aborted_queries = 0;
+  }
+
+  ~AC() { mysql_rwlock_destroy(&LOCK_ac); }
+  // Disable copy constructor.
+  AC(const AC &) = delete;
+  AC &operator=(const AC &) = delete;
+
+  inline void signal(std::shared_ptr<st_ac_node> &ac_node) {
```

Comment:
this is already inline
```suggestion
  void signal(st_ac_node_ptr &ac_node) {
```